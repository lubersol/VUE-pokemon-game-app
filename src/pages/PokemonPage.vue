<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else="pokemon">
    <h1>¿Quién es este pokemon?</h1>
    <pokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <pokemonOptions :pokemons="pokemonArr" @selectionPokemon="checkAnswer" />

    <div v-if="showAnswer">
      <h2 class="fade-in" :class="{ 'text-danger': hasError }">
        {{ message }}
      </h2>
      <button class="boton-jugar" @click="newGame">Jugar</button>
    </div>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOptions from "@/components/PokemonOptions";

import getPokemonOptions from "@/helpers/getPokemonOptions";

console.log(getPokemonOptions());

export default {
  components: { PokemonPicture, PokemonOptions },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
      hasError: false,
      "text-danger": false,
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      // Queremos un número random de los 4 elems del array.
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    //Emite un evento al hacer clik en el <li> y muestra la imagen en color
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (selectedId === this.pokemon.id) {
        this.message = `Correcto!, es ${this.pokemon.name}!`;
      } else {
        this.message = `Ooops!, era ${
          this.pokemon.name
        } ${(this.hasError = true)}`;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
      this.mixPokemonArray();
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style scoped>
.boton-jugar {
  background-color: #4caf50;
  border: 1px solid green;
  border-radius: 12px;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 16px;
  cursor: pointer;
  transition-duration: 0.4s;
}
.boton-jugar:hover {
  background-color: green;
  color: white;
}
.text-danger {
  color: red;
}
</style>
