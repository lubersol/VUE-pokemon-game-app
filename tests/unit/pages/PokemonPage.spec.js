import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', ()=>{
    let wrapper

    beforeEach(()=>{
        wrapper = shallowMount( PokemonPage )
    }) 

    test('debe de hacer match con el snapshot', () => {
        expect( wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar mixPokemonArray al montar', ()=>{

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando cargan los pokemons', ()=>{
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', ()=>{
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect(picture.exists() ).toBeTruthy()
        expect( options.exists() ).toBeTruthy()

        expect( picture.attributes('pokemonid') ).toBe('5')
        expect( options.attributes('pokemons') ).toBeTruthy()

    })

    test('pruebas con checkAnswer', async()=>{

        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(5)

        expect( wrapper.find('h2').exists()).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect( wrapper.find('h2').text() ).toBe(`Correcto, ${ pokemons[0].name }`)
  
        await wrapper.vm.checkAnswer(10)

        expect( wrapper.vm.message ).toBe(`Ooops!, era ${this.pokemon.name}`)
    })
})