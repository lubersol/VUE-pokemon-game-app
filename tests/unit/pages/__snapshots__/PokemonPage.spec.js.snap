// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
